### PR TITLE
rework cinder backend enablement

### PR DESCRIPTION
**Pull request by MaximilianMeister**

### CODE REVIEW

**Comment by aspiers** [comment]
Should the first two commits be squashed or are they really independent?  Also please can you give a bit more detail in those commit messages?  "rework" doesn't really tell the reviewer much :)

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -1814,13 +1804,13 @@ function set_proposalvars()
     fi
 
     # Cinder
-    if [[ ! $cinder_conf_volume_type ]] ; then
+    if [[ -z "$cinder_backend" ]] ; then
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: No need for -z and quotes here - its in double backets.
It was part of a consistency cleanup of this function.
I even thought about doing it for the entire script later.

**Comment by MaximilianMeister** [comment]
@jdsn done. further rebased, reordered the commits and enhanced the commit messages

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -1753,47 +1767,23 @@ function custom_configuration()
...
+            proposal_set_value cinder default "['attributes']['cinder']['volumes'][0]['backend_driver']" "'${cinder_backend}'"
+            case "$cinder_backend" in
+                netapp)
+                    cinder_netapp_proposal_configuration
+                    ;;
+            esac
 
-                # add a second backend to enable multi-backend, if not already present
+            # add a second backend to enable multi-backend, if not already present
+            if [ $want_cindermultibackend = 1 ] ; then
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: Either use double brackets or quotes here.

**Comment by MaximilianMeister** [comment]
strange github behaviour. i reordered the commits, by putting https://github.com/MaximilianMeister/automation/commit/a265b3e9b33071730ffbc437d741fcdd765f855c in front of https://github.com/MaximilianMeister/automation/commit/d3a5e63dd7ff1a2f9516ca268a300989206df4c8, and the github UI still shows the old order. the refs are correct though.

https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/ says

> If you always want to see commits in order, we recommend not using git rebase

So when somebody comments about reordering the commits, he needs to crawl through the refs before to be sure that they are in the wrong order...

**Comment by aspiers** [comment]
> So when somebody comments about reordering the commits, he needs to crawl through the refs before to be sure that they are in the wrong order...

No, he can just check out the branch in his local repo and view the real order there.

**Comment by bmwiedemann** [comment]
needs rebase

**Comment by MaximilianMeister** [comment]
rebased

File: scripts/mkcloud [hunk 3]
```diff
@@ -249,8 +250,9 @@ function sshrun()
 
         export localreposdir_target=$localreposdir_target ;
 
-        export cinder_conf_volume_type=$cinder_conf_volume_type;
-        export cinder_conf_volume_params=$cinder_conf_volume_params;
+        export cinder_backend=$cinder_backend;
+        export cinder_netapp_storage_protocol=$cinder_netapp_storage_protocol;
+        export want_cindermultibackend=$want_cindermultibackend;
```
Comment (bmwiedemann) [inline scripts/mkcloud]: want_ is already passed by the grep below

File: scripts/mkcloud [hunk 4]
```diff
@@ -1381,6 +1383,18 @@ function sanity_checks()
             complain 80 "Please set nodenumberlonelynode."
         fi
     fi
+
+    case $cinder_backend in
+        '')
+            :
+            ;;
+        netapp)
```
Comment (bmwiedemann) [inline scripts/mkcloud]: ''|netapp)

File: scripts/qa_crowbarsetup.sh [hunk 5]
```diff
@@ -1552,6 +1552,20 @@ function dns_proposal_configuration()
...
+function cinder_netapp_proposal_configuration()
+{
+    local attr_volume_path="['attributes']['cinder']['volumes']"
+    proposal_set_value cinder default "${attr_volume_path}[0]['backend_name']" "'netapp-backend'"
+    proposal_set_value cinder default "${attr_volume_path}[0]['netapp']['storage_family']" "'ontap_cluster'"
+    proposal_set_value cinder default "${attr_volume_path}[0]['netapp']['storage_protocol']" "'${cinder_netapp_storage_protocol}'"
+    proposal_set_value cinder default "${attr_volume_path}[0]['netapp']['netapp_server_hostname']" "'netapp-n1-e0m.cloud.suse.de'"
+    proposal_set_value cinder default "${attr_volume_path}[0]['netapp']['vserver']" "'cloud-openstack-svm'"
+    proposal_set_value cinder default "${attr_volume_path}[0]['netapp']['netapp_login']" "'openstack'"
+    proposal_set_value cinder default "${attr_volume_path}[0]['netapp']['netapp_password']" "'0penStack!'"
```
Comment (bmwiedemann) [inline scripts/qa_crowbarsetup.sh]: best use some variable, passed in from mkcloud, e.g. $cloud_netapp_password

**Comment by MaximilianMeister** [comment]
@bmwiedemann addressed your comments and tested it. please have another look

**Comment by bmwiedemann** [comment]
+0.5 from me. Code was introduced in 75bd26b3 by @toabctl - any comments on the parameter rework?
On one hand, having it all in one parameter makes it easier for passing it in through jenkins+mkcloud layers, but OTOH makes the parsing somewhat more unreadable...

**Comment by toabctl** [comment]
@MaximilianMeister what's the problem with the current configuration? Is there something you can't do currently?The commit mentioned by @bmwiedemann describes howto use the configuration. There was also a jenkins trigger for Cloud4+NetApp (no idea if that is still executed on a daily/weekly base)

**Comment by MaximilianMeister** [comment]
> Is there something you can't do currently?

no.

> what's the problem with the current configuration?

having keys and values in one parameter is not really good IMO. There is also no sanity check, and no documentation on how to use use it correctly. also i think this is the only mkcloud option which works like that. all the others work with just a single value.

I think with splitting it into clearly defined options is more user friendly and also the code, where the options are parsed is more readable in the end.

**Comment by jdsn** [comment]
:+1: lgtm

**Comment by jdsn** [comment]
@bmwiedemann @toabctl the issue with using it in mkcloud is IMHO no issue at all. The credentials are static, so they can go to the mkcloud job itself. The only variable we need in a trigger is the one that switches from one netapp to another or to another mode. The configuration variables are then set in the mkcloud job itself. In that case it does not matter how many variables we have to define.

@MaximilianMeister I agree that the code is more readable that way.

**Comment by toabctl** [comment]
@MaximilianMeister there is a sanity check - when applying the proposal. the parameters are mapped to the attributes and the proposal will fail if something is wrong.

File: scripts/mkcloud [hunk 6]
```diff
@@ -1381,6 +1386,14 @@ function sanity_checks()
             complain 80 "Please set nodenumberlonelynode."
         fi
     fi
+
+    case $cinder_backend in
+        ''|netapp|local|raw|rbd)
```
Comment (toabctl) [inline scripts/mkcloud]: I don't see how `local`, `raw` and `rbd` can be configured now. Or are the defaults fine?

File: scripts/mkcloud [hunk 7]
```diff
@@ -41,6 +40,10 @@ fi
 # causes of this is violation of the common shell coding standard which
 # uses uppercase for environment variables and constants, and lowercase
 # for local variables.
+: ${cinder_backend:=''}
+: ${cinder_netapp_storage_protocol:=iscsi}
+: ${cinder_netapp_login:=openstack}
+: ${cinder_netapp_password:=''}
```
Comment (toabctl) [inline scripts/mkcloud]: If we want to test more backends, the list of cinder_foo_bar variables could get very long. That's my only concern about this PR. But I'm not sure if that is a real problem.

Comment (toabctl) [inline scripts/mkcloud]: And another thing - what if we want to test netapp NFS and ISCSI together?

Comment (MaximilianMeister) [inline scripts/mkcloud]: > there is a sanity check - when applying the proposal. the parameters are mapped to the attributes and the proposal will fail if something is wrong.

I meant the sanity check in mkcloud, which would hit earlier, so you wouldn't have to wait for it so long to fail

> And another thing - what if we want to test netapp NFS and ISCSI together?

good point, we could do sth like:
``` bash
if [[ $want_cindermultibackend = 1 ]] ; then
    ...
    if [[ $cinder_backend = "netapp" ]]; then
        if [[ $cinder_netapp_storage_protocol = "iscsi" ]] ; then
            cinder_netapp_storage_protocol = "nfs"
        else
            cinder_netapp_storage_protocol = "iscsi"
        fi
        cinder_netapp_proposal_configuration
    fi
fi
```

I will have to adapt `cinder_netapp_proposal_configuration` as well to take the protocol into the `backend_name` that we can test both.
As discussed in with @jdsn I will open another PR for that after this one is merged

**Comment by MaximilianMeister** [comment]
@jdsn rebased

**Pull request merged**